### DISCUSSION TITLE
[FR] Randomize wg internal subnet address

### DISCUSSION THREAD

**Discussion by Champion0081** [pivpn/pivpn#1424] [Feature requests]
**Is your feature request related to a problem? Please describe.**
Currently pivpn sets the same internal wireguard subnet 10.6.0.1/24 for each installation and does not allow in the installer flow to change that subnet address. User unaware of that makes more than one install and the result is he cannot connect two tunnels to that two locations because of subnet conflict.

**Describe the solution you'd like**
randomize the subnet C part during installation 10.C.0.0 and make sure that is not conflicting with any subnets on current device interfaces or subnets in route table.

**Describe alternatives you've considered**
As there is no easy way to change this subnet after installation that needs to be addresed during installation. Tried editing pivpn vars and wg0.conf with no luck whatsoever. Tried those with no success: https://www.reddit.com/r/pivpn/comments/fysb4y/changing_the_pivpn_wireguard_default_network_space/
UPDATE possible via edition of files listed in this comment https://github.com/pivpn/pivpn/discussions/1424#discussioncomment-2272918

**Additional context**
Not needed, hope above is obvious.

**Comment by coolapso** [comment]
Hi Thanks for your Feature Request, just a note that there are some things you should have read and you did not: 

https://docs.pivpn.io/#general-guidelines

https://github.com/pivpn/pivpn#feedback--support

**Mainly**

https://github.com/pivpn/pivpn#feature-requests


I don't think the solution you propose is optimal ... 
But providing an **Advanced** option where more savvy users can tweak that setting sounds a little bit more sane. 

@orazioedoardo what you think? 



Moving this to discussions

**Reply by coolapso** [reply]
Its possible to do that with unattended setup anyway: https://github.com/pivpn/pivpn/blob/master/examples/unattended_wireguard_example.conf#L4  

I think I remember some discussion about this in the past to allow more savvy users to tweak those settings.

**Reply by Champion0081** [reply]
As i stated before i tried modyfying this variable but that does not work despite modifying the config on client side accordingly.
That is not only for savvy users this is for BEGINNERS. Currently we drag them into the trap of subnet conflict whenever they set up more than one location (my case). That is easy to implement and is a MUST imho. Please note that we even instruct in the installer flow on static/dhcp addresses issue, that is no less important.
PIVPN - as i understand - aims to be easy to set up. If we protect beginners from messing up the setup cause of DHCP, lets protect them from the subnet conflict too.
I googled thoroughly searching for easy chaning of the subnet address - found solutions did not work. If that is not easily searchable - that proves my point. Lets fix it several steps earlier - at installation.

**Reply by coolapso** [reply]
If its easy to implement ... i am glad to review your pull request ... thank you

**Reply by Champion0081** [reply]
@4s3ti  how that drastically differs in implementation from asking about static ip address during the installation? If you are not willing to fix that simple issue please pinpoint workaround - i couldnt find it and ways to change the subnet linked above did not work. Let's stop luring users into a trap by default.

**Reply by Champion0081** [reply]
It seems we just need  AskCustomNET ()
https://github.com/pivpn/pivpn/blob/master/auto_install/install.sh#L1347
https://github.com/pivpn/pivpn/blob/master/auto_install/install.sh#L1102

**Reply by coolapso** [reply]
@mattimac you are really starting with the wrong foot here!  Where have i said that I or WE PiVPN are not willing to fix anything?? Please mind the way you write! 

this being said 

**So here's your first answer:** 

https://docs.pivpn.io/#feedback-support

> PiVPN is purely community driven, and we are interested in making this script work for as many people as possible, we welcome any feedback on your experience. Please be respectful and be aware that this is maintained with our free time!

If you are willing to be part of the solution instead of the problem ... Man ... I will be more than glad to assist review your pull request! I am here for that ... as for serving your demands?? I'm better off using my time elsewhere Thank you! 

**here's your second answer:** 

https://github.com/pivpn/pivpn/issues/1273 and here https://docs.pivpn.io/install/#non-interactive-installation

Workaround as I said ... Install unattended and provide the subnet values that you wish on the Variables file. 

If you want to change it after installation ... the settings you are looking for are in `/etc/wireguard/wg0.conf` and `/etc/pivpn/wireguard/setupVars.conf`

**Reply by Champion0081** [reply]
@4s3ti  you completely misunderstood me. Please note "IF" instead of reacting emotionally and assuming the worst possible interpretation of the post. You have my feedback, i am not able currently to provide pull request for a couple of reasons. Let's at least try to understand the case and potentially discuss possible solutions for the case as i tried to unwrap why this is relevant. I respect your efforts and hope maybe we will draw some attention to the issue so that maybe someone else will be willing to contribute.
Thanks for fast replies

**Reply by coolapso** [reply]
This has been discussed before, and as I said on my original reply, I do not oppose at all to having an "advanced" option to change the VPN Subnet during the install, however might not jump into (my) priority list at the moment, however i am glad to review PR for this or any other feature/bug/etc..

One thing I do think we might be failing and might need some prioritization in the meantime is to document on how to change the subnet both with unattended or after the installation.

**Reply by TinCanTech** [reply]
Personally, I would prioritise in the reverse order.  Have the script allow for customising the `wireguard` subnet then document that change. That way, there is no need to spend time on documenting a work-around.

**Reply by coolapso** [reply]
@TinCanTech its not really documenting a workaround, what I ment is that there will always be someone needing to change the VPN Subnet post-install and that is the documentation that is indeed lacking


To sumarize: 

1 - We provide no whiptail dialog box for users to costumize the VPN Subnet (Only supported with unattended install)
2 - There is no documentation on how to change the Subnet Post-install

**Reply by TinCanTech** [reply]
@4s3ti I understand what you are saying, however .. honestly, I think it will be easier to allow customising the `WG` subnet in `pivpn` than it will be to document a post-install change.  Also, I believe users would rather trust `pivpn` to do it correctly for them, than to try to do it themselves.

I'll take a look myself.

**Reply by TinCanTech** [reply]
@4s3ti I had a look .. I see what you mean. I retract my previous comment. ;-)

**Reply by orazioedoardo** [reply]
My opinion on this topic:
Have the script automatically generate a random subnet by setting A and B in the 10.A.B.0/24 range, verify that there range does not collide with existing network interfaces (though there is a 1/65536 chance).

**Reply by Champion0081** [reply]
@orazioedoardo i would consider excluding known default subnets used to date for not only wireguard but other popular applications if possible just in case.

**Reply by orazioedoardo** [reply]
Which subnets in the 10.0.0.0/8 range other than 10.0.0.0/24 are defaults or commonly used?

**Reply by TinCanTech** [reply]
People have compiled this list, over time: https://community.openvpn.net/openvpn/wiki/AvoidRoutingConflicts

**Reply by orazioedoardo** [reply]
Merry Christmas, I've submitted the pull request: https://github.com/pivpn/pivpn/pull/1433

**Reply by coolapso** [reply]
PR #1433 has been merged to test branch.

**Reply by Champion0081** [reply]
> PR #1433 has been merged to test branch.

@orazioedoardo i want to clarify in reply to comment https://github.com/pivpn/pivpn/pull/1433#discussion_r775146017
Please note that the goal is not only to avoid routing conflicts on server side (current server where pivpn is getting installed) but the original issue resulted from conflict on *client* side (resulting from same wg interface subnet from different pivpn servers). That made maintaing multiple wg tunnels active to different pivpn wg servers impossible on desktop. 

As far as i understand PR #1433 addresses that by:
excluding BOTH:
 fixed list of subnets taken from above openvpn article
 list generated from interfaces on the host where pivpn is being installed

**Comment by Champion0081** [comment]
Hello, thank you for incorporating that into test branch. However still need 2 things:
1. How to change the subnet on existing installation without regenerating the configs. I assume there should be only changes required somewhere on the server and in the config. Lets assume current subnet is 10.6.0.0/24. How to change it on client and server to lets say 10.121.0.0/24 ?
2. How to install pivpn from test branch for dummies?

**Reply by orazioedoardo** [reply]
1. Not straightforward, you need to edit `/etc/pivpn/wireguard/setupVars.conf`, `/etc/wireguard/wg0.conf`, `/etc/iptables/rules.v4`.
2. https://docs.pivpn.io/install/#to-install-from-testdevelopment-branch

**Comment by Champion0081** [comment]
@orazioedoardo thank you for 1. I can confirm that works without reinstallation after changing the subnet in places listed and reboot, of course the subnet address need to be changed on client side too. Works flawlessly. Maybe we could incorporate information on that in documentation in case somebody needs that change before the change makes it to main branch?
